File: fluxstore-app/app/src/main/java/com/project/clothingstore/service/DashboardService.java
package com.project.clothingstore.service;

import com.google.firebase.Timestamp;
import com.google.firebase.firestore.CollectionReference;
import com.google.firebase.firestore.QueryDocumentSnapshot;
import com.project.clothingstore.utils.helper.FirebaseHelper;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

public class DashboardService {

    private static final CollectionReference userRef = FirebaseHelper.getUserCollection();
    private static final CollectionReference orderRef = FirebaseHelper.getOrderCollection();
    private static final CollectionReference ratingRef = FirebaseHelper.getRatingCollection();

    public void fetchDashboardData(Date selectedDate, ValueEventListener listener) {
        Timestamp startOfDay = getStartOfDay(selectedDate);
        Timestamp endOfDay = getEndOfDay(selectedDate);

        AtomicInteger orderCount = new AtomicInteger(0);
        AtomicInteger successCount = new AtomicInteger(0);

        orderRef
                .whereGreaterThanOrEqualTo("orderDate", startOfDay)
                .whereLessThanOrEqualTo("orderDate", endOfDay)
                .get()
                .addOnSuccessListener(orderQuery -> {
                    for (QueryDocumentSnapshot doc : orderQuery) {
                        orderCount.getAndIncrement();

                        String status = doc.getString("status");
                        if ("SUCCESS".equalsIgnoreCase(status)) {
                            successCount.getAndIncrement();
                        }
                    }

                    userRef
                            .whereGreaterThanOrEqualTo("createdAt", startOfDay)
                            .whereLessThanOrEqualTo("createdAt", endOfDay)
                            .get()
                            .addOnSuccessListener(userQuery -> {
                                int newUserCount = userQuery.size();

                                ratingRef
                                        .whereGreaterThanOrEqualTo("createdAt", startOfDay)
                                        .whereLessThanOrEqualTo("createdAt", endOfDay)
                                        .get()
                                        .addOnSuccessListener(ratingQuery -> {
                                            int ratingCount = ratingQuery.size();

                                            listener.onDataLoaded(
                                                    orderCount.get(),
                                                    newUserCount,
                                                    successCount.get(),
                                                    ratingCount
                                            );
                                        });
                            });
                });
    }

    private Timestamp getStartOfDay(Date date) {
        Calendar cal = Calendar.getInstance();
        cal.setTime(date);
        cal.set(Calendar.HOUR_OF_DAY, 0);
        cal.set(Calendar.MINUTE, 0);
        cal.set(Calendar.SECOND, 0);
        cal.set(Calendar.MILLISECOND, 0);
        return new Timestamp(cal.getTime());
    }

    private Timestamp getEndOfDay(Date date) {
        Calendar cal = Calendar.getInstance();
        cal.setTime(date);
        cal.set(Calendar.HOUR_OF_DAY, 23);
        cal.set(Calendar.MINUTE, 59);
        cal.set(Calendar.SECOND, 59);
        cal.set(Calendar.MILLISECOND, 999);
        return new Timestamp(cal.getTime());
    }
    public interface ValueEventListener {
        void onDataLoaded(int order, int newUser, int shipped, int rating);

        void onError(String error);
    }
    //Lấy dữ liệu thống kê 7 ngày gần nhất
    public void fetchRevenueLast7Days(OnRevenueDataListener listener) {
        Calendar calendar = Calendar.getInstance();
        calendar.set(Calendar.HOUR_OF_DAY, 0);
        calendar.set(Calendar.MINUTE, 0);
        calendar.set(Calendar.SECOND, 0);
        calendar.set(Calendar.MILLISECOND, 0);

        List<Timestamp> days = new ArrayList<>();
        for (int i = 6; i >= 0; i--) {
            Calendar cal = (Calendar) calendar.clone();
            cal.add(Calendar.DAY_OF_YEAR, -i);
            days.add(new Timestamp(cal.getTime()));
        }

        orderRef
                .whereGreaterThanOrEqualTo("orderDate", days.get(0))
                .get()
                .addOnSuccessListener(query -> {
                    Map<String, Double> revenueMap = new LinkedHashMap<>();

                    for (int i = 0; i < 7; i++) {
                        String key = formatDate(days.get(i).toDate());
                        revenueMap.put(key, 0.0);
                    }

                    for (QueryDocumentSnapshot doc : query) {
                        Timestamp orderDate = doc.getTimestamp("orderDate");
                        String status = doc.getString("status");
                        Double total = doc.getDouble("totalPrice");

                        if (orderDate != null && "SUCCESS".equalsIgnoreCase(status) && total != null) {
                            String key = formatDate(orderDate.toDate());
                            if (revenueMap.containsKey(key)) {
                                revenueMap.put(key, revenueMap.get(key) + total);
                            }
                        }
                    }

                    listener.onRevenueLoaded(revenueMap);
                });
    }

    private String formatDate(Date date) {
        SimpleDateFormat sdf = new SimpleDateFormat("dd/MM", Locale.getDefault());
        return sdf.format(date);
    }

    public interface OnRevenueDataListener {
        void onRevenueLoaded(Map<String, Double> dailyRevenue);

        void onError(String error);
    }

}

File: app/src/main/java/com/project/clothingstore/service/DashboardService.kt
package com.project.clothingstore.service

import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import com.project.clothingstore.modal.Order
import com.project.clothingstore.modal.RevenueEntry
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class DashboardService {

    private val db = FirebaseFirestore.getInstance()

    // Lấy tổng số đơn hàng
    fun getTotalOrders(onResult: (Int) -> Unit) {
        db.collection("orders").get()
            .addOnSuccessListener { snapshot ->
                onResult(snapshot.size())
            }
            .addOnFailureListener {
                onResult(0)
            }
    }

    // Lấy tổng doanh thu
    fun getTotalRevenue(onResult: (Float) -> Unit) {
        db.collection("orders").get()
            .addOnSuccessListener { snapshot ->
                var total = 0f
                for (doc in snapshot) {
                    val revenue = doc.getDouble("totalPrice") ?: 0.0
                    total += revenue.toFloat()
                }
                onResult(total)
            }
            .addOnFailureListener {
                onResult(0f)
            }
    }

    // Lấy tổng số người dùng
    fun getUserCount(onResult: (Int) -> Unit) {
        db.collection("users").get()
            .addOnSuccessListener { snapshot ->
                onResult(snapshot.size())
            }
            .addOnFailureListener {
                onResult(0)
            }
    }

    // Lấy số lượng sản phẩm còn trong kho
    fun getStockProductCount(onResult: (Int) -> Unit) {
        db.collection("products").get()
            .addOnSuccessListener { snapshot ->
                var count = 0
                for (doc in snapshot) {
                    val stock = doc.getLong("stock") ?: 0L
                    if (stock > 0) count++
                }
                onResult(count)
            }
            .addOnFailureListener {
                onResult(0)
            }
    }

    // Lấy các đơn hàng gần đây (5 đơn hàng gần nhất)
    fun getRecentOrders(onResult: (List<Order>) -> Unit) {
        db.collection("orders")
            .orderBy("createdAt", Query.Direction.DESCENDING)
            .limit(5)
            .get()
            .addOnSuccessListener { snapshot ->
                val orders = snapshot.mapNotNull { it.toObject(Order::class.java) }
                onResult(orders)
            }
            .addOnFailureListener {
                onResult(emptyList())
            }
    }

    // Lấy dữ liệu doanh thu theo tháng để vẽ biểu đồ
    fun getRevenueChartData(onResult: (List<RevenueEntry>) -> Unit) {
        db.collection("orders").get()
            .addOnSuccessListener { snapshot ->
                val map = mutableMapOf<String, Float>()
                for (doc in snapshot) {
                    val timestamp = doc.getTimestamp("createdAt")?.toDate()
                    val month =
                        SimpleDateFormat("MMM", Locale.getDefault()).format(timestamp ?: Date())
                    val revenue = doc.getDouble("totalPrice") ?: 0.0
                    map[month] = (map[month] ?: 0f) + revenue.toFloat()
                }
                val result = map.map { RevenueEntry(it.key, it.value) }
                onResult(result)
            }
            .addOnFailureListener {
                onResult(emptyList())
            }
    }
}

